File: Java/Containers/MongoDB-REST/src/main/java/com/api/mongodb/rest/service/mongodb/crud/MongoDB_CRUD_Client.java
package com.api.mongodb.rest.service.mongodb.crud;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;

import org.bson.BsonDocument;
import org.bson.conversions.Bson;
import org.bson.types.ObjectId;
import java.util.concurrent.CompletableFuture;

public class MongoDB_CRUD_Client implements IMongoDB_CRUD_Client {

    private final IMongoDB_Client_Builder _mongoDBClientBuilder;
    private MongoClient _mongoClient;

    public MongoDB_CRUD_Client(IMongoDB_Client_Builder mongoDBClientBuilder) {
        this._mongoDBClientBuilder = mongoDBClientBuilder;
    }

    private MongoClient getMongoClient() {
        if (_mongoClient == null) 
            _mongoClient = this._mongoDBClientBuilder.build();
        return _mongoClient;
    }

    private MongoCollection<BsonDocument> getCollection(String dbName, String collectionName) {
        MongoDatabase database = getMongoClient().getDatabase(dbName);
        return database.getCollection(collectionName, BsonDocument.class);
    }

    public CompletableFuture<Void> ping() {
        return CompletableFuture.runAsync(() -> {
            getMongoClient().getDatabase("admin").runCommand(new BsonDocument("ping", new BsonDocument()));
        });
    }

    public CompletableFuture<ObjectId> createAsync(String dbName, String collectionName, BsonDocument document) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            collection.insertOne(document);
            return document.getObjectId("_id").getValue();
        });
    }

    public CompletableFuture<BsonDocument> readAsync(String dbName, String collectionName, ObjectId documentId) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            Bson filter = Filters.eq("_id", documentId);
            return collection.find(filter).first();
        });
    }

    public CompletableFuture<BsonDocument> updateAsync(String dbName, String collectionName, ObjectId documentId, BsonDocument updateFields) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            Bson filter = Filters.eq("_id", documentId);
            BsonDocument update = new BsonDocument("$set", updateFields);
            return collection.findOneAndUpdate(filter, update);
        });
    }

    public CompletableFuture<Boolean> deleteAsync(String dbName, String collectionName, ObjectId documentId) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            Bson filter = Filters.eq("_id", documentId);
            return collection.deleteOne(filter).getDeletedCount() == 1;
        });
    }    
}

File: Examples/PHP/Containers/Laravel-MongoDB-REST/app/Services/MongoDB/CRUD/MongoDB_CRUD_Client.php
<?php


namespace App\Services\MongoDB\CRUD;

class MongoDB_CRUD_Client implements IMongoDB_CRUD_Client
{
    protected IMongoDB_Client_Builder $mongoClientBuilder;
    protected ?\MongoDB\Client $mongoClient;

    public function __construct(IMongoDB_Client_Builder $mongoClientBuilder)
    {
        $this->mongoClientBuilder = $mongoClientBuilder;
        $this->mongoClient = null;
    }

    private function getClient(): \MongoDB\Client
    {
        if ($this->mongoClient === null) {
            $this->mongoClient = $this->mongoClientBuilder->build();
        }
        return $this->mongoClient;
    }

    public function ping(): bool
    {
        $this->getClient()->admin->command(['ping' => 1]);
        return true;
    }

    protected function getCollection($dbName, $collectionName)
    {
        return $this->getClient()->$dbName->$collectionName;
    }

    public function createAsync(string $dbName, string $collectionName, \MongoDB\Model\BSONDocument $document): string
    {
        $collection = $this->getCollection($dbName, $collectionName);
        $result = $collection->insertOne($document);
        return $result->getInsertedId();
    }

    public function readAsync(string $dbName, string $collectionName, string $documentId): ?\MongoDB\Model\BSONDocument
    {
        $collection = $this->getCollection($dbName, $collectionName);
        return $collection->findOne(['_id' => new \MongoDB\BSON\ObjectId($documentId)]);
    }

    public function updateAsync(string $dbName, string $collectionName, string $documentId, \MongoDB\Model\BSONDocument $updateFields): ?\MongoDB\Model\BSONDocument
    {
        $collection = $this->getCollection($dbName, $collectionName);
        $update = ['$set' => $updateFields];
        return $collection->findOneAndUpdate(
            ['_id' => new \MongoDB\BSON\ObjectId($documentId)],
            $update,
            ['returnDocument' => \MongoDB\Operation\FindOneAndUpdate::RETURN_DOCUMENT_AFTER]
        );
    }

    public function deleteAsync(string $dbName, string $collectionName, string $documentId): bool
    {
        $collection = $this->getCollection($dbName, $collectionName);
        $result = $collection->deleteOne(['_id' => new \MongoDB\BSON\ObjectId($documentId)]);
        return $result->getDeletedCount() == 1;
    }
}